### PR TITLE
Better downloads/impressions cache logic

### PR DESCRIPTION
**Pull request by cavis**
🚧 WIP :crazy-idea:

This allows setting downloads/impressions cache keys in such a way that they:

1. Can always be fetched on-demand by Castle, in reasonable time, with the in-progress intervals expiring in 5 minutes
2. Can be INCR'd by an external lambda in a way to prevent Castle from recalculating them
3. Can be (eventually, not in this PR) recalculated by a Castle worker, at the top of every hour, to true-up the numbers

## Redis Caching

Previously, our cache keys looked something like:

```
downloads.podcast.70.HOUR.2017-10-19T08:00:00Z = 14841
downloads.podcast.70.HOUR.2017-10-19T09:00:00Z = 14832
downloads.episode.94b247a3-3d7c-4a07-97aa-95f87c539093.15MIN.2017-10-18T21:15:00 = 18505
```

Which for one, results in just a ton of cache keys.  Plus it makes it more difficult to redis-INCR them from an external lambda.  If an episode gets 0 downloads in a 15 minute interval ... the key never gets set, so it's a cache-miss, and Castle will go ask BigQuery for that interval (and get the 0 back).

This PR flips all cache keys into redis hashes:

```
downloads.podcasts.HOUR.2017-10-19T08:00:00Z = {70: 14841}
downloads.podcasts.HOUR.2017-10-19T09:00:00Z = {70: 14832}
downloads.episodes.15MIN.2017-10-18T21:15:00 = {94b247a3-3d7c-4a07-97aa-95f87c539093: 18505}
```

And castle is smart enough to know that _if a cache key exists_, any id not found in that hash means "0 downloads".  So with the above examples ... Castle would get a cache-hit (0) for podcast-id 99 at 9:00.  But it would get a cache-miss for podcast-id 99 at 10:00, and go ask BigQuery for that interval.

## Big Querying

Additionally, instead of asking BigQuery for a single podcast/episode, it will now ask for _all_ existing podcasts/episodes in that timeframe.  In theory, this is the exact same work/billing for BigQuery - since has a columnar structure, you're processing the same amount of data to get 1-episode vs 100-episodes.  But it does seem to take a bit longer to send that data over the wire, especially with episodes.  (Feeder is at about 3K episodes at the moment, fwiw).

## TODOs

In prod, we'll eventually have a lambda doing:

```
HINCRBY downloads.podcasts.HOUR.2017-10-19T08:00:00Z 70 1
EXPIRE downloads.podcasts.HOUR.2017-10-19T08:00:00Z 300
```

Probably tune that EXPIRE to happen at 15-minutes after the next hour.  Then 10-minutes after the hour, we'll run the Castle-worker to go through all keys about to expire, and actually calculate those values from BigQuery.

Thus: Castle download queries should never hit BigQuery.  And if they do, it's not a big deal.

Phew.

### CODE REVIEW

File: test/redis/interval_cache/getter_test.exs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    redis_clear("#{@prefix}*")
+    from = get_dtim("2017-03-22T01:15:00Z")
+    to = get_dtim("2017-03-22T02:15:00Z")
+    rollup = BigQuery.TimestampRollups.QuarterHourly
+    keys = Keys.keys(@prefix, rollup.range(from, to, false))
+    [from: from, to: to, rollup: rollup, keys: keys]
+  end
+
+  test "hits the entire range", %{from: from, to: to, rollup: rollup, keys: keys} do
+    Conn.hset(Enum.at(keys, 0), "key1", 11)
```
Comment (gcampo88) [inline test/redis/interval_cache/getter_test.exs]: might vote for one of these `key`s being renamed? maybe `keys` could be named `cached_intervals` or something? or `key1` etc could be `podcast1`? a little confusing at first glance otherwise

File: test/redis/interval_cache/getter_test.exs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+      with_mock Timex, [:passthrough], [now: fn() -> get_dtim(time) end] do
+        {hits, _new_from} = Getter.get(@prefix, "key1", from, to, rollup)
+        {length(hits), List.last(hits).count}
+      end
+    end
+    assert hits_at_time.("2017-03-22T03:00:00Z") == {3, 33}
+    assert hits_at_time.("2017-03-22T02:00:00Z") == {4, 0}
+    assert hits_at_time.("2017-03-22T02:00:29Z") == {4, 0}
+    assert hits_at_time.("2017-03-22T02:00:31Z") == {3, 33}
+    assert hits_at_time.("2017-03-22T01:00:00Z") == {4, 0}
```
Comment (gcampo88) [inline test/redis/interval_cache/getter_test.exs]: not quite sure what's going on in last one (line 103)? is it assuming that since the later 3 intervals have values, that that one didn't get set, and returns 0 for that interval?

File: test/redis/interval_cache/setter_test.exs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      assert Setter.interval_ttl(get_dtim("2017-03-22T01:15:00Z")) == 300
+      assert Setter.interval_ttl(get_dtim("2017-03-22T02:20:00Z")) == 300
+    end
+  end
+
+  test "sets empty counts", %{from: from, to: to} do
+    assert redis_count("#{@prefix}*") == 0
+    Setter.set(@prefix, from, to, %{})
+    assert redis_count("#{@prefix}*") == 1
+    assert Conn.command(["TTL", List.first(redis_keys("#{@prefix}*"))]) == -1
```
Comment (gcampo88) [inline test/redis/interval_cache/setter_test.exs]: so this last line is checking that the setter didn't put a TTL on the empty count, right?

File: test/redis/interval_cache_test.exs [hunk 4]
```diff
@@ -1,192 +1,92 @@
...
-    assert meta.cache_hits == 3
-
-    {data, meta} = interval @prefix, from, to, QuarterHourly, fn(_new_from) ->
-      raise "should not have called this"
-    end
-    assert Enum.map(data, &(&1.count)) == [55, 66, 77, 88, 99, 111]
-    assert format_dtim(hd(data).time) == "2017-03-22T01:15:00Z"
-    assert meta.cached == true
-    assert meta.cache_hits == 6
+  test "partially hits the cache", %{interval: intv, keys: keys} do
```
Comment (gcampo88) [inline test/redis/interval_cache_test.exs]: a little puzzled by this test -- it's testing that the attempt to set the same key (key #3) with a different value (44 instead of 43) won't work, right? i'm not sure i understand why `meta.cache_hits` is 2 and not 3 though, if 0, 1, and 3 are all set? i think i'm missing something...

File: test/support/time_helpers.ex [hunk 5]
```diff
@@ -0,0 +1,25 @@
+defmodule Castle.TimeHelpers do
+  defmacro __using__(_opts) do
```
Comment (gcampo88) [inline test/support/time_helpers.ex]: cooollll

File: lib/big_query/base/timestamp.ex [hunk 6]
```diff
@@ -1,51 +1,57 @@
...
-  def timestamp_sql(tbl, interval, where_sql) do
-    """
-    WITH intervals AS (#{timestamp_intervals(tbl, interval, where_sql)})
-    SELECT time, count FROM intervals
-    ORDER BY time ASC
+  def timestamp_sql(tbl, interval, group_by) do
     """
+    WITH intervals AS (#{timestamp_intervals(tbl, interval, group_by)})
+    SELECT time, #{group_by}, count FROM intervals
+    ORDER BY time ASC, #{group_by} ASC
```
Comment (gcampo88) [inline lib/big_query/base/timestamp.ex]: won't this leave you with `ORDER BY time ASC,  ASC` if group_by ain't there?

Comment (cavis) [reply lib/big_query/base/timestamp.ex]: It should always be there.  There's no longer a `timestamp_sql/2` with a nil 3rd param ... and nothing from timestamp_group chains to this.

File: test/redis/interval_cache/getter_test.exs (hunk 1 above)
Comment (cavis) [reply test/redis/interval_cache/getter_test.exs]: +yes

File: test/redis/interval_cache/getter_test.exs (hunk 2 above)
Comment (cavis) [reply test/redis/interval_cache/getter_test.exs]: Yeah, so this is testing that we won't ask BQ for data about `2:00 - 2:15` until a few seconds after that interval starts.  There won't be any data there anyways, until kinesis/lambda have processed it.  (I have it set to 30 seconds right now, so it waits until `2:00:30` to ask).

The last line is an edge case test ... if it's `1:00` now, and you're asking for data an hour in the future ... I'm just going to assume it's 0.  Because how could BQ know the future?

File: test/redis/interval_cache/setter_test.exs (hunk 3 above)
Comment (cavis) [reply test/redis/interval_cache/setter_test.exs]: It's actually checking that it didn't set a TTL, because `to` is more than `@current_interval_buffer` (1 hour) in the past.  And after an hour, we can assume that kinesis/bigquery have processed any impressions/downloads for that interval, and just cache the result from BQ forever.

File: test/redis/interval_cache_test.exs (hunk 4 above)
Comment (cavis) [reply test/redis/interval_cache_test.exs]: Since key-2 is a miss, it doesn't even check the cache for key-3.  It asks BQ for both 2 and 3, then assembles the result `[0-cached, 1-cached, 2-from-BQ, 3-from-BQ]`.

File: test/support/time_helpers.ex (hunk 5 above)
Comment (cavis) [reply test/support/time_helpers.ex]: Underscores are cool.  The more, the cooler.

File: test/redis/interval_cache_test.exs (hunk 4 above)
Comment (gcampo88) [reply test/redis/interval_cache_test.exs]: ah, because key-2 would be present if it had been queried before, even if everything in it was 0. so if it misses one key it assumes nothing after it is present either. 👍

File: test/support/time_helpers.ex (hunk 5 above)
Comment (gcampo88) [reply test/support/time_helpers.ex]: that is my understanding of underscores as well.
